pkg/network/ovn_kubernetes.go

```diff
@@ -334,3 +372,53 @@ func listenDualStack(masterIP string) string {
 		return ""
 	}
 }
+
+// shouldUpdateOvnKMaster determines if we're ready to roll out
+// upgrades to the master daemonset.
+//
+// We shouldn't upgrade ovn-k master until all the nodes are upgraded, because of
+// compatability questions
+func shouldUpdateMaster(existingNode, existingMaster *appsv1.DaemonSet) bool {
+	releaseVersion := os.Getenv("RELEASE_VERSION")
+	// Fresh cluster - full steam ahead!
+	if existingNode == nil || existingMaster == nil {
+		return true
+	}
+
+	nodeVersion := existingNode.GetAnnotations()["release.openshift.io/version"]
+
+	if nodeVersion != releaseVersion {
+		// TODO: expose this directly as status. Need to refactor status-manager
+		klog.V(3).Infof("Not upgrading ovn-kube master yet, ovn-kube node DaemonSet still at version %s, need %s",
+			nodeVersion, releaseVersion)
+		return false
+	}
+
+	status := existingNode.Status
+
+	// Copy-pasted from status_manager: Figure out if a given DaemonSet is
+	// not fully rolled out.
+	nodeProgressing := (status.UpdatedNumberScheduled < status.DesiredNumberScheduled ||
+		status.NumberUnavailable > 0 ||
+		status.NumberAvailable == 0 ||
+		existingNode.Generation > status.ObservedGeneration)
+
+	// If progressing, don't upgrade masters yet.
+	// Exception: a rollout hung at 90% is overrided.
+	nodeHung := existingNode.GetAnnotations()[names.RolloutHungAnnotation] == "true"
```

Comment:
I am still not convinced about this, but I am willing to let it go. The concern I have is: if a roll-out is hung at X % (meaning it's made no progress at all for `ProgressTimeout` = 10 min), it's probably never going to make it to 90 %, and it feels like we should be forcing it in those cases too, not only when it's close to the finish line. Do you have any concern with just checking for the annotation being true, and forcing the roll-out in that case?